Cargo.toml

```diff
@@ -410,7 +410,18 @@ xcm-procedural = { git = "https://github.com/subspace/polkadot-sdk", rev = "ad62
 # De-duplicate extra copy that comes from Substrate repo
 substrate-bip39 = "0.6.0"
 
+[patch."https://github.com/autonomys/rust-libp2p.git"]
+# Patch away `libp2p` in our dependency tree with the git version.
+# This brings the fixes in our `libp2p` fork into substrate's dependencies.
+#
+# This is a hack: patches to the same repository are rejected by `cargo`. But it considers
+# "subspace/rust-libp2p" and "autonomys/rust-libp2p" to be different repositories, even though
+# they're redirected to the same place by GitHub, so it allows this patch.
+libp2p = { git = "https://github.com/subspace/rust-libp2p", rev = "399c4c7fcba821a7bac2663e090a7b155b08ebe1" }
+libp2p-identity = { git = "https://github.com/subspace/rust-libp2p", rev = "399c4c7fcba821a7bac2663e090a7b155b08ebe1" }
+multistream-select = { git = "https://github.com/subspace/rust-libp2p", rev = "399c4c7fcba821a7bac2663e090a7b155b08ebe1" }
```

Comment:
This looks really awkward. I understand why it works, but is it really not possible to patch `autonomys` with a different `autonomys` revision? Or wouldn't it converge on the same revision automatically on its own?

Reply:
No, patches can only come from different sources, because `[patch]` is a source-to-source replacement, filtered by crate name:
https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-patch-section

It's not super clear in the documentation, but `cargo` returns the error:
> patch for `libp2p` in `https://github.com/autonomys/subspace` points to the same source, but patches must point to different sources

https://github.com/rust-lang/cargo/blob/ab15d58608e1d6895fae4ab9a400efe49a75a13f/src/cargo/core/registry.rs#L441

And it does canonicalisation, which means we can't just use `subspace.git` and `subspace`.

Reply:
We also can't convince it to converge on the same revision, because that only works for `crates.io` and similar registries. Git versions are fixed to the specific commit hash.

Reply:
Very unfortunate :confused: